File: src/pytypes/onefuzztypes/requests.py
#!/usr/bin/env python
#
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import AnyHttpUrl, BaseModel, Field, root_validator

from ._monkeypatch import _check_hotfix
from .consts import ONE_HOUR, SEVEN_DAYS
from .enums import (
    OS,
    Architecture,
    JobState,
    NodeState,
    PoolState,
    ScalesetState,
    TaskState,
)
from .events import EventType
from .models import AutoScaleConfig, InstanceConfig, NotificationConfig
from .primitives import Container, PoolName, Region
from .webhooks import WebhookMessageFormat


class BaseRequest(BaseModel):
    pass


class JobGet(BaseRequest):
    job_id: UUID


class JobSearch(BaseRequest):
    job_id: Optional[UUID]
    state: Optional[List[JobState]]
    task_state: Optional[List[TaskState]]
    with_tasks: Optional[bool]


class NotificationCreate(BaseRequest, NotificationConfig):
    container: Container
    replace_existing: bool = Field(default=False)


class NotificationSearch(BaseRequest):
    container: Optional[List[Container]]


class NotificationGet(BaseRequest):
    notification_id: UUID


class TaskGet(BaseRequest):
    task_id: UUID


class TaskSearch(BaseRequest):
    job_id: Optional[UUID]
    task_id: Optional[UUID]
    state: Optional[List[TaskState]]


class TaskResize(TaskGet):
    count: int = Field(ge=1)


class NodeCommandGet(BaseRequest):
    machine_id: UUID


class NodeCommandDelete(BaseRequest):
    machine_id: UUID
    message_id: str


class AgentRegistrationGet(BaseRequest):
    machine_id: UUID


class AgentRegistrationPost(BaseRequest):
    pool_name: PoolName
    scaleset_id: Optional[UUID]
    machine_id: UUID
    version: str = Field(default="1.0.0")


class PoolCreate(BaseRequest):
    name: PoolName
    os: OS
    arch: Architecture
    managed: bool
    client_id: Optional[UUID]
    autoscale: Optional[AutoScaleConfig]


class PoolSearch(BaseRequest):
    pool_id: Optional[UUID]
    name: Optional[PoolName]
    state: Optional[List[PoolState]]


class PoolStop(BaseRequest):
    name: PoolName
    now: bool


class ProxyGet(BaseRequest):
    scaleset_id: Optional[UUID]
    machine_id: Optional[UUID]
    dst_port: Optional[int]

    @root_validator()
    def check_proxy_get(cls, value: Any) -> Any:
        check_keys = ["scaleset_id", "machine_id", "dst_port"]
        included = [x in value for x in check_keys]
        if any(included) and not all(included):
            raise ValueError(
                "ProxyGet must provide all or none of the following: %s"
                % ", ".join(check_keys)
            )
        return value


class ProxyCreate(BaseRequest):
    scaleset_id: UUID
    machine_id: UUID
    dst_port: int
    duration: int = Field(ge=ONE_HOUR, le=SEVEN_DAYS)


class ProxyDelete(BaseRequest):
    scaleset_id: UUID
    machine_id: UUID
    dst_port: Optional[int]


class NodeSearch(BaseRequest):
    machine_id: Optional[UUID]
    state: Optional[List[NodeState]]
    scaleset_id: Optional[UUID]
    pool_name: Optional[PoolName]


class NodeGet(BaseRequest):
    machine_id: UUID


class NodeUpdate(BaseRequest):
    machine_id: UUID
    debug_keep_node: Optional[bool]


class ScalesetSearch(BaseRequest):
    scaleset_id: Optional[UUID]
    state: Optional[List[ScalesetState]]
    include_auth: bool = Field(default=False)


class ScalesetStop(BaseRequest):
    scaleset_id: UUID
    now: bool


class ScalesetUpdate(BaseRequest):
    scaleset_id: UUID
    size: Optional[int] = Field(ge=1)


class AutoScaleOptions(BaseModel):
    min: int = Field(ge=0)
    max: int = Field(ge=1)
    default: int = Field(ge=0)
    scale_out_amount: int = Field(ge=1)
    scale_out_cooldown: int = Field(ge=1)
    scale_in_amount: int = Field(ge=1)
    scale_in_cooldown: int = Field(ge=1)


class ScalesetCreate(BaseRequest):
    pool_name: PoolName
    vm_sku: str
    image: Optional[str]
    region: Optional[Region]
    size: int = Field(ge=1)
    spot_instances: bool
    ephemeral_os_disks: bool = Field(default=False)
    tags: Dict[str, str]
    auto_scale: Optional[AutoScaleOptions]


class ContainerGet(BaseRequest):
    name: Container


class ContainerCreate(BaseRequest):
    name: Container
    metadata: Optional[Dict[str, str]]


class ContainerDelete(BaseRequest):
    name: Container
    metadata: Optional[Dict[str, str]]


class ReproGet(BaseRequest):
    vm_id: Optional[UUID]


class ProxyReset(BaseRequest):
    region: Region


class CanScheduleRequest(BaseRequest):
    machine_id: UUID
    task_id: UUID


class WebhookCreate(BaseRequest):
    name: str
    url: AnyHttpUrl
    event_types: List[EventType]
    secret_token: Optional[str]
    message_format: Optional[WebhookMessageFormat]


class WebhookSearch(BaseModel):
    webhook_id: Optional[UUID]


class WebhookGet(BaseModel):
    webhook_id: UUID


class WebhookUpdate(BaseModel):
    webhook_id: UUID
    name: Optional[str]
    event_types: Optional[List[EventType]]
    url: Optional[AnyHttpUrl]
    secret_token: Optional[str]
    message_format: Optional[WebhookMessageFormat]


class NodeAddSshKey(BaseModel):
    machine_id: UUID
    public_key: str


class InstanceConfigUpdate(BaseModel):
    config: InstanceConfig


_check_hotfix()

File: src/ApiService/ApiService/OneFuzzTypes/Requests.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Microsoft.OneFuzz.Service;

public record BaseRequest {
    [JsonExtensionData]
    public Dictionary<string, JsonElement>? ExtensionData { get; set; }
};

public record CanScheduleRequest(
    [property: Required] Guid MachineId,
    [property: Required] Guid TaskId
) : BaseRequest;

public record NodeCommandGet(
    [property: Required] Guid MachineId
) : BaseRequest;

public record NodeCommandDelete(
    [property: Required] Guid MachineId,
    [property: Required] string MessageId
) : BaseRequest;

public record NodeGet(
    [property: Required] Guid MachineId
) : BaseRequest;

public record NodeUpdate(
    [property: Required] Guid MachineId,
    bool? DebugKeepNode
) : BaseRequest;

public record NodeSearch(
    Guid? MachineId = null,
    List<NodeState>? State = null,
    ScalesetId? ScalesetId = null,
    PoolName? PoolName = null
) : BaseRequest;

public record NodeStateEnvelope(
    [property: Required] NodeEventBase Event,
    [property: Required] Guid MachineId
) : BaseRequest;

// either NodeEvent or WorkerEvent
[JsonConverter(typeof(SubclassConverter<NodeEventBase>))]
public abstract record NodeEventBase;

public record NodeEvent(
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    NodeStateUpdate? StateUpdate,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    WorkerEvent? WorkerEvent
) : NodeEventBase;

public record WorkerEvent(
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    WorkerDoneEvent? Done = null,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    WorkerRunningEvent? Running = null
) : NodeEventBase;

public record WorkerRunningEvent(
    [property: Required] Guid TaskId);

public record WorkerDoneEvent(
    [property: Required] Guid TaskId,
    [property: Required] ExitStatus ExitStatus,
    [property: Required] string Stderr,
    [property: Required] string Stdout);

public record NodeStateUpdate(
    [property: Required] NodeState State,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    NodeStateData? Data = null
) : NodeEventBase;

// NodeSettingUpEventData, NodeDoneEventData, or ProcessOutput
[JsonConverter(typeof(SubclassConverter<NodeStateData>))]
public abstract record NodeStateData;

public record NodeSettingUpEventData(
   [property: Required] List<Guid> Tasks
) : NodeStateData;

public record NodeDoneEventData(
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    string? Error,

    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    ProcessOutput? ScriptOutput
) : NodeStateData;

public record ProcessOutput(
    ExitStatus ExitStatus,
    string Stderr,
    string Stdout
) : NodeStateData;

public record ExitStatus(
    int? Code,
    int? Signal,
    bool Success);

public record ContainerGet(
    [property: Required] Container Name
) : BaseRequest;

public record ContainerCreate(
    [property: Required] Container Name,
    IDictionary<string, string>? Metadata = null
) : BaseRequest;

public record ContainerDelete(
    [property: Required] Container Name,
    IDictionary<string, string>? Metadata = null
) : BaseRequest;

public record NotificationCreate(
    [property: Required] Container Container,
    [property: Required] bool ReplaceExisting,
    [property: Required] NotificationTemplate Config
) : BaseRequest;

public record NotificationSearch(
    List<Container>? Container,
    Guid? NotificationId
) : BaseRequest;


public record NotificationTest(
    [property: Required] IReport Report,
    [property: Required] Notification Notification
) : BaseRequest;

public record NotificationGet(
    [property: Required] Guid NotificationId
) : BaseRequest;

public record JobGet(
    [property: Required] Guid JobId
) : BaseRequest;

public record JobCreate(
    [property: Required] string Project,
    [property: Required] string Name,
    [property: Required] string Build,
    [property: Required] long Duration,
    string? Logs
) : BaseRequest;

public record JobSearch(
    Guid? JobId = null,
    List<JobState>? State = null,
    List<TaskState>? TaskState = null,
    bool? WithTasks = null
) : BaseRequest;

public record NodeAddSshKeyPost(
    [property: Required] Guid MachineId,
    [property: Required] string PublicKey
) : BaseRequest;

public record ReproGet(Guid? VmId) : BaseRequest;

public record ReproCreate(
    [property: Required] Container Container,
    [property: Required] string Path,
    [property: Required] long Duration
) : BaseRequest;

public record ProxyGet(
    ScalesetId? ScalesetId,
    Guid? MachineId,
    int? DstPort
) : BaseRequest;

public record ProxyCreate(
    [property: Required] ScalesetId ScalesetId,
    [property: Required] Guid MachineId,
    [property: Required] int DstPort,
    [property: Required] int Duration
) : BaseRequest;

public record ProxyDelete(
    [property: Required] ScalesetId ScalesetId,
    [property: Required] Guid MachineId,
    int? DstPort
) : BaseRequest;

public record ProxyReset(
    [property: Required] string Region
) : BaseRequest;

public record ScalesetCreate(
    [property: Required] PoolName PoolName,
    [property: Required] string VmSku,
    ImageReference? Image,
    Region? Region,
    [property: Range(1, long.MaxValue), Required] long Size,
    [property: Required] bool SpotInstances,
    [property: Required] Dictionary<string, string> Tags,
    bool EphemeralOsDisks = false,
    AutoScaleOptions? AutoScale = null
) : BaseRequest;

public record AutoScaleOptions(
    [property: Range(0, long.MaxValue)] long Min,
    [property: Range(1, long.MaxValue)] long Max,
    [property: Range(0, long.MaxValue)] long Default,
    [property: Range(1, long.MaxValue)] long ScaleOutAmount,
    [property: Range(1, long.MaxValue)] long ScaleOutCooldown,
    [property: Range(1, long.MaxValue)] long ScaleInAmount,
    [property: Range(1, long.MaxValue)] long ScaleInCooldown
);

public record ScalesetSearch(
    ScalesetId? ScalesetId = null,
    List<ScalesetState>? State = null,
    bool IncludeAuth = false
) : BaseRequest;

public record ScalesetStop(
    [property: Required] ScalesetId ScalesetId,
    [property: Required] bool Now
) : BaseRequest;

public record ScalesetUpdate(
    [property: Required] ScalesetId ScalesetId,
    [property: Range(1, long.MaxValue)]
    long? Size
) : BaseRequest;

public record TaskGet(
    [property: Required] Guid TaskId
) : BaseRequest;

public record TaskCreate(
   [property: Required] Guid JobId,
   List<Guid>? PrereqTasks,
   [property: Required] TaskDetails Task,
   [property: Required] TaskPool Pool,
   List<TaskContainers>? Containers = null,
   Dictionary<string, string>? Tags = null,
   List<TaskDebugFlag>? Debug = null,
   bool? Colocate = null
) : BaseRequest;

public record TaskSearch(
    Guid? JobId,
    Guid? TaskId,
    List<TaskState>? State = null
) : BaseRequest;

public record PoolSearch(
    Guid? PoolId = null,
    PoolName? Name = null,
    List<PoolState>? State = null
) : BaseRequest;

public record PoolStop(
    [property: Required] PoolName Name,
    [property: Required] bool Now
) : BaseRequest;

public record PoolCreate(
    [property: Required] PoolName Name,
    [property: Required] Os Os,
    [property: Required] Architecture Arch,
    [property: Required] bool Managed,
    Guid? ObjectId = null,
    bool Update = false
) : BaseRequest;


public record PoolUpdate(
    [property: Required] PoolName Name,
    Guid? ObjectId = null
) : BaseRequest;

public record WebhookCreate(
    [property: Required] string Name,
    [property: Required] Uri Url,
    [property: Required] List<EventType> EventTypes,
    string? SecretToken,
    WebhookMessageFormat? MessageFormat
) : BaseRequest;

public record WebhookSearch(Guid? WebhookId) : BaseRequest;

public record WebhookGet([property: Required] Guid WebhookId) : BaseRequest;

public record WebhookUpdate(
    [property: Required] Guid WebhookId,
    string? Name,
    Uri? Url,
    List<EventType>? EventTypes,
    string? SecretToken,
    WebhookMessageFormat? MessageFormat
) : BaseRequest;

public record InstanceConfigUpdate(
    [property: Required] InstanceConfig config
) : BaseRequest;


public record AgentRegistrationGet(
    [property: Required] Guid MachineId
) : BaseRequest;


public record AgentRegistrationPost(
    [property: Required] PoolName PoolName,
    ScalesetId? ScalesetId,
    [property: Required] Guid MachineId,
    Os? Os,
    string? MachineName,
    [property: Required] string Version = "1.0.0"
) : BaseRequest;

public record TemplateValidationPost(
    [property: Required] string Template,
    TemplateRenderContext? Context
) : BaseRequest;

public record JinjaToScribanMigrationPost(
    bool DryRun = false
) : BaseRequest;

public record EventsGet(
    [property: Required] Guid EventId
) : BaseRequest;
